Readerlink: Make it so buttons can be added beyond module

roygbyte opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/9746)
<!-- Reviewable:end -->

poire-z commented:
Looks like you got reverted a few of the things that has been changed to this file (probably the 8 latest commits at https://github.com/koreader/koreader/commits/master/frontend/apps/reader/modules/readerlink.lua).

roygbyte commented:
Right you are :) There, I think I got it right this time.

poire-z commented:
> Right you are :) There, I think I got it right this time.

Wrong you think :)
There are still things not related to your button stuff in the github diff (location_stack, math.pow, FOLLOW_LINK_TIMEOUT...)

roygbyte commented:
Oh boy. Can I just mail in my changes?

roygbyte commented:
How about thiiiiis?

poire-z commented on frontend/apps/reader/modules/readerlink.lua:
```diff
@@ -688,62 +689,50 @@ end
 
 function ReaderLink:onGoToExternalLink(link_url)
     local text
```
You're populating `self.alt_handlers_buttons` in here (previously, we were creating a `local alt_handlers_buttons` ), so on each external link, you're trashing and recreating most of the keys. This should probably better dont in :init(), like ReaderHighlight:init() does.

roygbyte replied:
I think the menu gets trashed and recreated because the options that are populated will vary according to the type of external link. For instance: if the link is a wiki link, the menu will show a wiki option. Maybe there's a clever way to keep it efficient, but my PR is just concerned with being able to add more buttons to the menu from outside of the module

poire-z replied:
> my PR is just concerned with being able to add more buttons to the menu from outside of the module

I get it, but there must be a way to do this more cleanly :)
Either continue to get the table local, and only register your external buttons in a self.something - and add them to the local.
Or (my preference) build & keep in :init() a self.all_buttons with each a `should_add(link_url)` and/or `is_enabled(link_url)` functions, and have your plugin add them to that. Then in this function, just call these methods to build the local button table.

poire-z commented on frontend/apps/reader/modules/readerlink.lua:
```diff
@@ -22,7 +22,8 @@ local Screen = Device.screen
 local ReaderLink = InputContainer:extend{
-    location_stack = nil, -- table, per-instance
+    location_stack = nil, -- table, per-instance,
+    alt_handlers_buttons = {}
```
Should be per-instance, in :init().

NiLuJe replied:
This is still an issue in the latest version ;).

NiLuJe replied:
And you'll want to possibly make it clearer whether it should be provided by the constructor table, or, more likely given the latest changes, at runtime via the added API ;).

poire-z commented on frontend/apps/reader/modules/readerlink.lua:
```diff
@@ -1405,4 +1399,32 @@ function ReaderLink:showAsFootnotePopup(link, neglect_current_location)
     return true
 end
 
+function ReaderLink:addButton(idx, fn_button)
+    self.alt_handlers_buttons[idx] = fn_button
+end
+
+function ReaderLink:removeButton(idx)
```
I think I mentionned that in your previous PR (or one of them :), but you'd better try to keep the same interface/API as ReaderHighlight, which already allows plugins to add buttons (although none currently use it).

Curious: which of your plugins will add what buttons ?

roygbyte replied:
``` lua
function ReaderHighlight:addToHighlightDialog(idx, fn_button)
    -- fn_button is a function that takes the ReaderHighlight instance
    -- as argument, and returns a table describing the button to be added.
    self._highlight_buttons[idx] = fn_button
end
```

It's fairly similar. I can def change the method name to be something like `addToExternalLinkDialog`.

I have a new plugin that will download a link to an EPUB. That would add a single button to "Download to EPUB". (Often, I'm reading articles from the RSS reader, and they contain links, and I want to read the links! So hence the "download to EPUB" :)  )

Frenzie commented:
Search in Wikipedia, save as EPUB, it'll be full of 'em.

(Or just make an HTML file with `<a href="https://en.wikipedia.org/wiki/Wikipedia">Wikipedia</a>`.)

roygbyte commented:
> Search in Wikipedia, save as EPUB, it'll be full of 'em.

Thanks, I had forgotten that was a feature.

Ok--I'm really hoping this is good enough. Definitely agree with @poire-z's reasoning about moving the buttons to `init` and having a function to call to determine if the button should display. Makes the function for generating the dialog much cleaner.

poire-z commented:
You need to clean your changes a bit (or best, report/redo your real changes on a clean readerlink.lua), there's lot of indentation changes that make it hard to see if you tweaked this code or not.
Also, if you make it so plugins can add buttons, there is a candidate in there that shall go away in its own plugin: wallabag.

roygbyte commented:
Thanks for the help! I've reverted the indents and moved the Wallabag button. Thought about doing the same for wiki button, but that's not a plugin... so probably fine to keep in `readerlink`?

poire-z commented on frontend/apps/reader/modules/readerlink.lua:
```diff
@@ -91,6 +92,138 @@ function ReaderLink:init()
+            text = _("Show QR code"),
+            callback = function()
+                UIManager:close(this.dialog)
+                UIManager:show(QRMessage:new{
+                        text = link_url,
```
You really like to mess our nice "4 spaces per level" indentation :/ Or it is while moving it around and fixing manually each line indentation (and your text editor is not a good friend).
(Others below).

roygbyte replied:
I'm sorry :( I'm using Emacs and I don't really know how to do setup up the automatic indents properly.

poire-z commented on frontend/apps/reader/modules/readerlink.lua:
```diff
@@ -91,6 +92,138 @@ function ReaderLink:init()
+                UIManager:close(this.dialog)
+                Device:openLink(link_url)
+            end,
+            show_in_dialog_func = function()
+                if Device:canOpenLink() then
+                    return true
+                end
```
I was going to say you could just use `show_in_dialog_func = Device.canOpenLink` as it is itself a straight function.
But I remember @NiLuJe may have said something about using them as methods recently ? (I may have imagining that :/)
Anyway, I don't mind the above, I find it more readable.

roygbyte replied:
`Device.canOpenLink` was returning some weird value. It wasn't boolean, from what I could tell. So I had to do this.

NiLuJe replied:
If there's no argument (and as such no risk of confusion on what the "object" in `self` is, an alias should work?

roygbyte replied:
Not sure how to do whatever you're suggesting :)

poire-z replied:
Just `show_in_dialog_func = Device.canOpenLink,` (try it to see if it works).
But the current is just fine.

poire-z commented on frontend/apps/reader/modules/readerlink.lua:
```diff
@@ -91,6 +92,138 @@ function ReaderLink:init()
+            end
+        }
+    end
+    self._external_buttons["07_cancel"] = function(this, link_url)
```
Where would you place your future buttons?
We had some discussions with the highlight dialog buttons, as depending on stuff, we may get an even or an odd number of buttons, and so the last one may be wide (when even number). We decided for "Search" to be that last.
For external links, "Cancel" is obviously the candidate, so no real problem there :)

But may be number the names better ? 10_copy ...  90_cancel so your plugin can more easily insert itself between these ?

poire-z commented on frontend/apps/reader/modules/readerlink.lua:
```diff
@@ -91,6 +92,138 @@ function ReaderLink:init()
+            end
+        }
+    end
+    local find_epub_fullpath = function(link_url)
```
This could be a module local function, instead of being fabricated on each :init() - although it would be a bit far from here were it is used (but it would make this section more readable).
Anyway, mention "wikipedia" in its name: `find_wikipedia_epub_fullpath()`

Or may be put all the logic and regexp in a single function:
`is_wiki_page(link_uri, check_if_epub_exists) => wiki_page, wiki_lang, epub_fullpath`
which may make this section smaller ?

roygbyte replied:
Great suggestion, thanks!

poire-z commented on frontend/apps/reader/modules/readerlink.lua:
```diff
@@ -687,137 +820,30 @@ function ReaderLink:onGotoLink(link, neglect_current_location, allow_footnote_po
+                    -- The default title is replaced by the first non-default button title.
+                    title = button_title
+                else
+                    -- Every other button title value is appended to the title.
+                    title = title .. "\n\n" .. button_title
```
(We'll see if this can happen, and how it works in practice, we might need to find wording that look fine together.)

roygbyte replied:
In practice, the titles are appended when there's a wiki page lookup for an existing wiki page. No other usage that I know of.

poire-z commented on frontend/apps/reader/modules/readerlink.lua:
```diff
@@ -687,137 +820,30 @@ function ReaderLink:onGotoLink(link, neglect_current_location, allow_footnote_po
-    dialog = ButtonDialogTitle:new{
-        title = text,
-        buttons = buttons,
+    self.dialog = ButtonDialogTitle:new{
```
Name this `self.external_link_dialog`, it makes it easier to grasp when reading all the little `function(this, link_uri)` where it's already not intuitive to get what `this` is.

roygbyte replied:
`self.dialog` is used elsewhere in this module, and I'm not sure I want to disturb or make sense of those usages... In those contexts, `self.dialog` actually isn't always an external link (In once place it's a Footnote widget).

poire-z replied:
Oh, but that's not the same!
`self.dialog` is the parent of readerlink, so readerui, and it is used when refreshing (in many widgets, we pass the real parent widget (the widget UIManager:show()n) down to each little widget that makes this real widget via self.dialog. So, when a button needs to refresh itself, it requests self.dialog to repaint/refresh this little region (so, we know we don't need to repaint/refresh all the other widgets, ie the book, the footer).

So, one reason to really give it another name (previously, it was a `local dialog`, nothing to do with `self.dialog`.

roygbyte commented on frontend/apps/reader/modules/readerlink.lua:
```diff
@@ -23,6 +23,7 @@ local T = ffiutil.template
 
 local ReaderLink = InputContainer:extend{
     location_stack = nil, -- table, per-instance
+    _external_buttons = {}
```
Oh you know what... I did have this variable initialized in init. But I found that the buttons added from elsewhere got overwritten by the init. Should I use two tables? One for default buttons (initialized in ReaderLink), and a second for other buttons?

NiLuJe replied:
That sounds like either a code or a design issue ;).

NiLuJe replied:
The name of the variable implies that this should only contain extra buttons already, which means the code actually building the widgets shouldn't solely rely on it?

poire-z replied:
> But I found that the buttons added from elsewhere got overwritten by the init

That might be because you are/were using ReaderLink:addToExternalLinkDialog instead of self.ui.link:addToExternalLinkDialog , so adding them to the class object instead of the instance (ie. in wallabag, 2 comments below this one).

poire-z replied:
> The name of the variable implies that this should only contain extra buttons already

Depends on how you read that :) I am reading it as "external_link_buttons" - not as "external_external_link_buttons" :)

roygbyte replied:
Nevermind. I can just leave it nil and init the table when the first button is added.

roygbyte commented on plugins/wallabag.koplugin/main.lua:
```diff
@@ -113,6 +114,19 @@ function Wallabag:init()
         self.is_dateparser_checked = true
     end
+
+    ReaderLink:addToExternalLinkDialog("03_wallabag", function(this, link_url)
```
I don't understand this. I revised it, as suggested, but get this error when launching: ` Failed to initialize wallabag plugin: plugins/wallabag.koplugin/main.lua:118: attempt to index field 'link' (a nil value) `

poire-z replied:
OK, this plugin is enabled also when not in Reader (ie starting with file manager?). So:
```lua
if self.ui and self.ui.link then
    self.ui.link:addToExternalLinkDialog(...) 
end 
```
?

roygbyte replied:
Well I don't understand this, but it works, so that's good enough!

poire-z replied:
If you're in Filemanager, the wallbag plugin is loaded. self.ui is the apps/filemanager.lua instance, and it does not have an attribute named `link`.
If you're in Reader, the wallbag plugin is loaded. self.ui is the apps/reader.lua instance, and it does have an attribute named `link`: the ReaderLink instance.

NiLuJe replied:
Much like widgets, there's only *one* plugin class, the one initially loaded by `require` (we'll gloss over how pluginloader does this for plugins, as it does not actually use require, because the end result is the same).

On the other hand, much like widgets, you *will* have multiple *instances* of plugins: every time ReaderUI or FileManager starts, *every* active plugins gets a new instance (i.e., `init` gets called). Conversely, when one of them is torn down, the plugins follow suit (much like widgets, they get a `CloseWidget` event).

Like ReaderUI/FileManager modules (whose lifecycle is similar), plugins inherit a pointer to their parent via self.ui. That will either point to the current ReaderUI or FileManager instance, so, if you're using fields/methods specific to only one of them, you need to be careful, as @poire-z explained ;).

roygbyte commented on frontend/apps/reader/modules/readerlink.lua:
```diff
@@ -91,6 +92,93 @@ function ReaderLink:init()
+            callback = function()
+                UIManager:scheduleIn(0.1, function()
+                    UIManager:close(this.dialog)
+                    local _,_ wiki_epub_fullpath = is_wiki_page(link_url)
```
I didn't test it. But even with the syntax mistake, it doesn't crash! But it doesn't open, either _, _

poire-z replied:
`local _,_, wiki_epub_fullpath`. You forgot a space in between

poire-z commented:
Don't forget/miss my comment about self.dialog above.

melyux commented:
@roygbyte Could you tell me what this PR does?

roygbyte commented:
> @roygbyte Could you tell me what this PR does?

It makes it so buttons can be added to ReaderLink's external links dialog box. When you click a link in a text, this dialog box appears. I want to create a plugin that adds buttons to the box. Hence, this PR :)

poire-z commented on frontend/apps/reader/modules/readerlink.lua:
```diff
@@ -687,137 +775,30 @@ function ReaderLink:onGotoLink(link, neglect_current_location, allow_footnote_po
-                local epub_path = dir .. "/" .. safe_filename
-                if util.pathExists(epub_path) then
-                    epub_fullpath = epub_path
+    local default_title =  T(_("External link:\n\n%1"), BD.url(link_url))
+    local title = default_title
+
+    for _, fn_button in pairs(self._external_link_buttons) do
+        local button = fn_button(self, link_url)
+        if button.show_in_dialog_func then
+            local show, button_title = button.show_in_dialog_func()
```
You will be calling each button.show_in_dialog_func() twice:
- once here, to know if it provides a title, and build the final `title`
- once below with self:getButtonsForExternalLinkDialog
```lua
    self.external_link_dialog = ButtonDialogTitle:new{
        title = title,
        buttons = self:getButtonsForExternalLinkDialog(link_url),
    }
```
May be your self:getButtonsForExternalLinkDialog() should itself build and return the title? 
`local buttons, title = self:getButtonsForExternalLinkDialog(link_url)`

roygbyte replied:
Excellent (: ty

roygbyte commented on plugins/wallabag.koplugin/main.lua:
```diff
@@ -18,6 +18,7 @@ local Math = require("optmath")
 local MultiConfirmBox = require("ui/widget/multiconfirmbox")
 local MultiInputDialog = require("ui/widget/multiinputdialog")
 local NetworkMgr = require("ui/network/manager")
+local ReaderLink = require("apps/reader/modules/readerlink")
```
`is_wiki_page` is def used, so that's weird. I moved the function to the top of the file, which tho inelegant should take care of the CI issue?

poire-z replied:
Yes, as it is a module local function, it needs to be defined before it is used.
You could just add a `local is_wiki_page` at the top - so it "exists" for the users after, and define it later `is_wiki_page = function()... end` - but that's also inelegant. So best to define it earlier, we have such stuff for example in readerdictionary.lua.

poire-z commented:
```
Luacheck results
Checking frontend/apps/reader/modules/readerlink.lua 5 warnings

    frontend/apps/reader/modules/readerlink.lua:139:50: accessing undefined variable 'is_wiki_page'
    frontend/apps/reader/modules/readerlink.lua:144:46: accessing undefined variable 'is_wiki_page'
    frontend/apps/reader/modules/readerlink.lua:161:54: accessing undefined variable 'is_wiki_page'
    frontend/apps/reader/modules/readerlink.lua:166:66: accessing undefined variable 'is_wiki_page'
    frontend/apps/reader/modules/readerlink.lua:1422:10: setting non-standard global variable 'is_wiki_page'
```

roygbyte commented:
Hmmm... `is_wiki_page` is defined at the end of the file. Should I move it above? Everything seems to work the way it is, tho. Suggestions?

roygbyte commented:
Ahhh! Where would I be without you, @poire-z  ^^
